tcp-server/src/main.rs

```diff
@@ -0,0 +1,116 @@
+use std::borrow::BorrowMut;
+use std::collections::vec_deque::Iter;
+use std::collections::VecDeque;
+use std::io;
+use std::io::{BufRead, BufReader, Read, Write};
+use std::net::{TcpListener, TcpStream};
+use std::ops::Deref;
+use std::sync::{Arc, Mutex};
+use std::thread;
+use std::thread::JoinHandle;
+use std::time::Duration;
+
+use redisish::{Command, parse};
+
+trait Mailbox {
+    /// TODO Is it fine that mutability is hidden? How can we do it some other way?
+    fn append(&self, email: String);
+    /// TODO does it make sense to return an abstract iterator and how can this be achieved?
+    fn list_emails(&self) -> String;
+}
+
+struct VecDequeMailbox {
+    data: Mutex<VecDeque<String>>,
+}
+
+impl Mailbox for VecDequeMailbox {
+    /// TODO is there a way (and does it make sense) to accept &str?
+    fn append(&self, email: String) {
+        let result = self.data.lock();
+        result.unwrap().push_front(email);
+    }
+
+    /// Returns a list of emails as a string, separated by `;`
+    fn list_emails(&self) -> String {
+        self.data.lock()
+            .unwrap()
+            .iter()
+            .fold(String::new(), |acc, next| acc + next + ";")
+    }
+}
+
+impl VecDequeMailbox {
+    // TODO how to return Box<dyn Mailbox>?
```

Comment:
you have only one implementation, so I would not use traits + dyn here; generally speaking you should be able to just declare `Box<dyn Mailbox>` as return type. Ping me if you have issues with that!

Reply:
Going with async functions means that Mailbox cannot be a trait for now. :(
My motivation was to make Mailbox an abstraction for multiple implementations,  for example, VecDequeMailbox, FileMailbox, FirebaseMailbox.